### DISCUSSION TITLE
Does TEDIT-SEE work with {NODIRCORE} files?

### DISCUSSION THREAD

**Discussion by pamoroso** [Interlisp/medley#1235] [General]
Suppose I create a temporary core file:

```lisp
(SETQ S (OPENSTREAM '{NODIRCORE} 'OUTPUT}
(PRINTOUT S "This is some text" T)
(FORCEOUTPUT S)
```

If I try to display the file with `TEDIT-SEE`:

```lisp
(TEDIT-SEE S)
```

I get the error `File not found: #.(PATHNAME "")`. Is `TEDIT-SEE` supposed to work with `{NODIRCORE}` files? Am I going through the right steps? According to IRM:

> The "name" of such a file is simply the stream object returned from `(OPENSTREAM ’{NODIRCORE} ’OUTPUT)`, and it is this stream object that must be passed to all input/output operations, including `CLOSEF` and any calls to `OPENSTREAM` to reopen the file.

So I think the stream `S` may be passed to `TEDIT-SEE`.

`TEDIT-SEE` works fine with `{CORE}` files.

**Comment by rmkaplan** [comment]
One issue is that TEDIT (and TEDIT-SEE) want to open the file for input, but the stream is specified as OUTPUT in this example.

But if you open it BOTH, it still fails.  But the failure is not in TEDIT-SEE or TEDIT, it seems that CL:OPEN doesn’t work right for {NODIRCORE}:

For
(SETQ X (CL:OPEN ‘{NODIRCORE} :IO))

you get a pathname that shows an empty string, and then

(CL:OPEN X :INPUT) gives the file-not-found error.

TEDIT-SEE (but not TEDIT) runs into this because TEDIT-SEE is using CL:WITH-OPEN-FILE to guarantee that the input file will be closed on exit from this scope, and CL:WITH-OPEN-FILE uses CL:OPEN.  (TEDIT-SEE will actually translate a Lisp source file into a scratch image file so that the font-change and other possible formatting control characters are rendered properly.)

So I think the bug is in the way that CL:OPEN treats {NODIRCORE} files, possibly because those files by definition don’t have ordinary directory-like file names.  Or possibly the fix would be to make the NODIRCORE take better care to fake up what looks like an actual name that would make the commonlisp functions happy.

Separately, it doesn’t make sense to do TEDIT-SEE on an empty input stream.

**Reply by nbriggs** [reply]
Perhaps TEDIT-SEE it should be using CL:WITH-OPEN-STREAM, which evaluates the expression to produce a stream, instead of CL:WITH-OPEN-FILE which evaluates the expression to produce a file-spec?  Though it still doesn't make sense to pass a {NODIRCORE} file to it, since either way it will be closed when complete.

**Comment by pamoroso** [comment]
Thanks, I wasn't sure this was a bug or some kind of expected behavior. This discussion may be turned into a issue, but using `{CORE}` files is a good workaround.

**Comment by pamoroso** [comment]
> Separately, it doesn’t make sense to do TEDIT-SEE on an empty input stream.

@rmkaplan What do you mean?

**Comment by rmkaplan** [comment]
The difference is that the storage for {NODIRCORE} gets collected when all references to the stream are gone, whereas {CORE} files will say around until they are explicitly deleted (or the sysout is discarded).

**Comment by masinter** [comment]
A string stream would be simpler, wouldn't it?

**Reply by pamoroso** [reply]
If you mean passing as the first argument of `TEDIT-SEE` a string stream opened with `OPENSTRINGSTREAM`, I still get the `File not found: #.(PATHNAME "")` error.

**Reply by masinter** [reply]
suggest converting this discussion to an issue, tag: Bug, project: system

**Comment by rmkaplan** [comment]
I have fixed TEDIT-SEE in the new TEDIT, so that it uses OPENSTREAM instead of the commonlisp macros, and therefore does OK on stream arguments. And, if FILE is an already open stream (like {NODIRCORE}), it leaves the file open on exit.

But that doesn’t address the more general problem that the commonlisp macros are more sensitive than Interlisp function with respect to stream arguments.  That may still be an issue.

**Reply by pamoroso** [reply]
I haven't had a chance to test this yet as the latest Interlisp Online was built in April of 2023 and it doesn't incorporate the `{NODIRCORE}` fix. But, in general, calling `TEDIT` directly may be a good workaround as it does support `{NODIRCORE}`.

**Reply by fghalasz** [reply]
@pamoroso   I just upgraded Interlisp Online with latest Medley release (medley-230816-59bc12ff).

-- Frank

**Reply by pamoroso** [reply]
Thanks, I've just tried the 15-Sep-2023 23:14:52 Interlisp Online build and I still get the error `File not found: #.(PATHNAME "")` when passing a `{NODIRCORE}` stream as the first argument of `TEDIT-SEE`.

**Reply by fghalasz** [reply]
@pamoroso It looks like no TEDIT code has been merged into master since last year.  All of Ron's recent work on TEDIT is being kept in a separate branch on Github.  So none of the recent TEDIT work is included in the official release builds.  And since Interlisp Online only runs official release builds, none of the recent TEDIT changes will make it to online until Ron's branch(es) are merged into master.

Perhaps this is a discussion topic for Monday meeting.

**Reply by pamoroso** [reply]
Makes sense, got it.
